src/d3a/models/area/event_dispatcher.py

```diff
@@ -73,34 +82,33 @@ def broadcast_balancing_market_cycle(self, **kwargs):
     def broadcast_callback(self):
         return self._broadcast_notification
 
-    def _broadcast_notification(self, event_type: Union[MarketEvent, AreaEvent], **kwargs):
-        if not self.area.events.is_enabled and \
-           event_type not in [AreaEvent.ACTIVATE, AreaEvent.MARKET_CYCLE]:
-            return
-        # Broadcast to children in random order to ensure fairness
-        for child in sorted(self.area.children, key=lambda _: random()):
-            child.dispatcher.event_listener(event_type, **kwargs)
-        # Also broadcast to IAAs. Again in random order
-        for time_slot, agents in self._inter_area_agents.items():
-            if time_slot not in self.area._markets.markets:
+    def _broadcast_notification_to_agents_of_market_type(
+            self, market_type: AvailableMarketTypes, event_type: AreaEvent, **kwargs):
+        for time_slot, agents in self._get_agents_for_market_type(self, market_type).items():
+            if time_slot not in self.area._markets.get_market_instances_from_class_type(
+                    market_type):
                 # exclude past IAAs
                 continue
 
             if not self.area.events.is_connected:
                 break
             for area_name in sorted(agents, key=lambda _: random()):
                 agents[area_name].event_listener(event_type, **kwargs)
-        # Also broadcast to BAs. Again in random order
-        # TODO: Refactor to reuse the spot market mechanism
-        for time_slot, agents in self._balancing_agents.items():
-            if time_slot not in self.area._markets.balancing_markets:
-                # exclude past BAs
-                continue
 
-            if not self.area.events.is_connected:
-                break
-            for area_name in sorted(agents, key=lambda _: random()):
-                agents[area_name].event_listener(event_type, **kwargs)
+    def _broadcast_notification(self, event_type: Union[MarketEvent, AreaEvent], **kwargs):
+        if not self.area.events.is_enabled and \
+           event_type not in [AreaEvent.ACTIVATE, AreaEvent.MARKET_CYCLE]:
+            return
+
+        # Broadcast to children in random order to ensure fairness
+        for child in sorted(self.area.children, key=lambda _: random()):
+            child.dispatcher.event_listener(event_type, **kwargs)
+        self._broadcast_notification_to_agents_of_market_type(
+            AvailableMarketTypes.SPOT, event_type, **kwargs)
+        self._broadcast_notification_to_agents_of_market_type(
+            AvailableMarketTypes.BALANCING, event_type, **kwargs)
+        self._broadcast_notification_to_agents_of_market_type(
+            AvailableMarketTypes.SETTLEMENT, event_type, **kwargs)
```

Comment:
Nit: in this class we have many `if` methods to basically distinguish between SPOT, BALANCING and SETTLEMENT market types (like [here](https://github.com/gridsingularity/d3a/blob/ea445ce2d33609c69a7dcd577f6978113475b137/src/d3a/models/area/event_dispatcher.py#L165-L174) and [here](https://github.com/gridsingularity/d3a/blob/ea445ce2d33609c69a7dcd577f6978113475b137/src/d3a/models/area/event_dispatcher.py#L144-L163)). 
I think that this could be refactored into separate classes (each specialized on one market). No need to do it in the context of this PR, but it might be good to keep it in mind and maybe open a refactoring ticket.

Reply:
I would leave this for when we reimplement the event dispatcher to not forward the bids and the offers from the higher to the lower market, since this is going to be implemented in the near future (this is the last task of the degrees of freedom market)